Added context manager enabling all futures returned by rpc_async and custom build rpc functions to be automatically waited on

osandoval-fb opened the pull request:
Test Plan: Make sure ci tests pass, including newly written test

Differential Revision: D22640839

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -136,6 +136,16 @@ def _set_proceed_shutdown_signal(sequence_id):
+thread_local_var = threading.local()
+
+@contextlib.contextmanager
+def wait_all():
```
For public APIs, we will need to add API docs for it. You can refer to other public APIs in this file (e.g., `shutdown`) as examples. The rString in `shutdown()` API compiles to this API doc: https://pytorch.org/docs/master/rpc.html#torch.distributed.rpc.shutdown

mrshenli replied:
To build docs locally and verify results, you can follow this steps https://github.com/pytorch/pytorch#building-the-documentation. It might be a little hard to do it in fbcode (but it's possible, let me know if you want to try that path). You can, however,  clone PyTorch repo to your laptop, and build PyTorch and the html doc files locally, and then use [this script](https://gist.github.com/colesbury/df218a96334dbd31ee0e93760f70f114) to serve it.

osandoval-fb replied:
<img width="1413" alt="wait_all_documentation" src="https://user-images.githubusercontent.com/68608730/89812093-9808e900-db0d-11ea-8c00-3891fc26c3e2.png">

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -136,6 +136,16 @@ def _set_proceed_shutdown_signal(sequence_id):
     ), "Termination signal sequence id {} got set twice.".format(sequence_id)
     proceed_signal.set()
 
+thread_local_var = threading.local()
```
We don't expect users to access this var, so let's add a "_" prefix to its name.

osandoval-fb replied:
@mrshenli If I add a _prefix the test doesnt have access to rpc._thread_local_var, and it needs that.  Is there a preferred way around this limitation?  If not then the name needs to be thread_local_var

mrshenli replied:
accessing private vars in our own tests are fine, we just don't expect users to use it. Other private vars like `_internal_rpc_pickler` are also accessed in test. Adding a `_` prefix is more like a convention to politely tell users "please don't access this var", but nothing actually prevent them from doing so.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2731,6 +2731,16 @@ class TestPickler:
+        with rpc.wait_all():
+            self.assertTrue(rpc.thread_local_var.future_list == [])
+            dst = worker_name((self.rank + 1) % self.world_size)
+            fut = rpc.rpc_async(dst, torch.add, (torch.ones(2, 2), 1))
```
Let's add some more tests:

0. Run multiple `rpc_sync` and `rpc_async` calls in one context. 
1. the RPC target function throws an error
2. the RPC target function timed out

mrshenli reviewed (changes requested):
Thanks for adding this! Code looks good to me. Could you please rebase to the latest master to get rid of irrelevant test failures?

osandoval-fb commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2731,6 +2731,16 @@ class TestPickler:
             torch.distributed.rpc.api._default_pickler is _internal_rpc_pickler
         )
 
+    @dist_init
+    def test_wait_all(self):
+        with rpc.wait_all():
+            self.assertTrue(rpc.thread_local_var.future_list == [])
+            dst = worker_name((self.rank + 1) % self.world_size)
+            fut = rpc.rpc_async(dst, torch.add, (torch.ones(2, 2), 1))
```
@mrshenli I presume it would be preferred to check that the only element in future_list is a C futures object here.  If you agree, how would I do that?

mrshenli replied:
You can try `isinstance(rpc.thread_local_var.future_list[0], torch._C.Future)`

osandoval-fb commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2731,6 +2731,16 @@ class TestPickler:
+            dst = worker_name((self.rank + 1) % self.world_size)
+            fut = rpc.rpc_async(dst, torch.add, (torch.ones(2, 2), 1))
+            self.assertTrue(len(rpc.thread_local_var.future_list) == 1)
+            self.assertEqual(fut.wait(), torch.ones(2, 2) + 1)
```
@mrshenli  I'm confused why we need this, seems independent of newly created function we are testing here.  It does however seem reasonable to, after the with block is over, test whether the future in future_list was waited by the context manager.  Is there a way to do this?

mrshenli replied:
Yep, you can access `fut` outside of the with context, as Python var lifetime is not scoped by context managers.

osandoval-fb replied:
@mrshenli But then is
``` self.assertEqual(fut.wait(), torch.ones(2, 2) + 1)```
actually needed or may I delete it?

mrshenli replied:
Let's keep this, just to make sure `rpc_async` still returns the correct future object.

osandoval-fb commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2731,6 +2731,49 @@ class TestPickler:
+                self.assertTrue(rpc.thread_local_var.future_list == [])
+                dst = worker_name((self.rank + 1) % self.world_size)
+                fut = rpc.rpc_async(dst, raise_func)
+        self.assertTrue(len(rpc.thread_local_var.future_list) == 1)
```
@mrshenli  Note if an exception is raised by any future in future_list that is waited on, and that exception is caught in the application, future_list will not be deleted.  Ie:

`try:
	with rpc.wait_all():
		fut = rpc.rpc_async(dst, torch.add, (torch.ones(2, 2), 1))
		fut = rpc.rpc_async(dst, raise_func)
except BaseException:
	#rpc.thread_local_var.future_list still exists with 2 futures in it`

Since future_list is reinitialized as [] each time wait_all Is invoked, this should be fine?  On the other hand we could catch any error raised by torch.futures.wait_all and delete future_list before passing the error up, but the user may want access to the future_list in the event of an exception.  What do you think?

mrshenli replied:
>  On the other hand we could catch any error raised by torch.futures.wait_all and delete future_list before passing the error up, but the user may want access to the future_list in the event of an exception. What do you think?

Good call! We can add a `collect_all()` API (similar to [the one in `torch.futures`](https://pytorch.org/docs/master/futures.html#torch.futures.collect_all)) for this purpose. I mean sth like:

```python
with rpc.collect_all() as futs:
   # make rpc calls

   # access futs here if necessary
```

For `wait_all()` I'd vote for letting it delete `future_list` and  re-throw the first exception. Open to other suggestions. :)

mrshenli replied:
We can add `collect_all()` in a followup PR.

osandoval-fb commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2731,6 +2737,61 @@ class TestPickler:
+            fut = rpc.rpc_async(dst, torch.add, (torch.ones(2, 2), 1))
+            was_waited_ptr = WasFutureWaitedHelper()
+            print("original was waited ptr type is {0}".format(was_waited_ptr))
+            rpc.rpc_async(dst, set_arg_to_waited, args=(was_waited_ptr,))
```
@mrshenli Logically set_arg_to_waited should be updating was_waited_ptr before the corresponding assert statement but this is not happening; its as if set_arg_to_waited is receiving a copy of was_waited_ptr and updating that instead.  

I wanted to ask first do you have a preferred way in mind of testing whether the context manager successfully waited the future?  If not do you off the top of your head know why my expected behavior is not occurring?

If not no worries I'll find the solution, I just don't want to waste more time debugging this if you had something else in mind

mrshenli replied:
> I wanted to ask first do you have a preferred way in mind of testing whether the context manager successfully waited the future?

Good point, let me add a `done()` API to future.

> If not do you off the top of your head know why my expected behavior is not occurring?

Because they are operating on different objects. When you pass `was_waited_ptr` to `rpc_async`, the `was_waited_ptr` will be pickled on the caller, and then unpickled on the callee, and then the callee will set it's `was_future_waited` field. So the `was_waited_ptr` would keep intact.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -312,6 +312,12 @@ def raise_func():
 
+class WasFutureWaitedHelper():
+    def __init__(self):
+        self.was_future_waited = False
```
let's use two blank lines to separate top-level vars and functions.

osandoval-fb committed:
Added context manager enabling all futures returned by rpc_async and custom build rpc functions to be automatically waited on (#41807)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/41807

Test Plan: Make sure ci tests pass, including newly written test

Differential Revision: D22640839

fbshipit-source-id: 6c38df048de5619c4308e555b241f482f6f82d23

osandoval-fb committed:
Added context manager enabling all futures returned by rpc_async and custom build rpc functions to be automatically waited on (#41807)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/41807

Test Plan: Make sure ci tests pass, including newly written test

Differential Revision: D22640839

fbshipit-source-id: 6c38df048de5619c4308e555b241f482f6f82d23

osandoval-fb committed:
added wait all autofunction to rpc.rst

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -134,6 +134,34 @@ def _broadcast_to_followers(sequence_id, objects_map):
+@contextlib.contextmanager
+def wait_all():
+    r"""
+    A context manager that collects all futures returned by rpc_async and
```
double quote `rpc_async` to make it a code block

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2741,6 +2741,59 @@ class TestPickler:
+            self.assertTrue(_thread_local_var.future_list == [])
+            dst = worker_name((self.rank + 1) % self.world_size)
+            fut = rpc.rpc_async(dst, torch.add, (torch.ones(2, 2), 1))
+            self.assertFalse(fut.done())
```
this can potentially be flaky, in case the `fut` is done before the check? Maybe remove this?

mrshenli reviewed (commented):
LGTM! Thanks for adding this feature with extensive tests. And thanks for checking the doc build. Since this is phase 1 of the feature, let's not expose its API doc to PyTorch website just yet (by removing `autofunction`) to avoid API behavior churn. We can expose it when we are able to capture all futures returned by `RpcAgent::send` (this will however break the current `rpc_sync`-related tests). The doc build exercise is still valuable as we are now confident that the doc string is correct.

osandoval-fb committed:
final change responses

osandoval-fb commented:
@mrshenli Assuming there aren't any other changes we want, what does the rest of the process look like to get this landed?  Do I need to squash commits?

mrshenli reviewed (approved):
LGTM! Can you rebase to master to clear the merge conflict and test errors. Then you can land through the phabricator.

mrshenli reviewed (changes requested):
Test error are real on the latest version:

```
Aug 18 00:15:04 ======================================================================
Aug 18 00:15:04 ERROR [1.220s]: test_wait_all_multiple_call (__main__.ProcessGroupRpcTestWithSpawn)
Aug 18 00:15:04 ----------------------------------------------------------------------
Aug 18 00:15:04 Traceback (most recent call last):
Aug 18 00:15:04   File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 223, in wrapper
Aug 18 00:15:04     self._join_processes(fn)
Aug 18 00:15:04   File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 330, in _join_processes
Aug 18 00:15:04     self._check_return_codes(elapsed_time)
Aug 18 00:15:04   File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 363, in _check_return_codes
Aug 18 00:15:04     raise RuntimeError(error)
Aug 18 00:15:04 RuntimeError: Processes 0 1 2 3 exited with error code 10
```

osandoval-fb committed:
fixed test

osandoval-fb committed:
fixed test 2

osandoval-fb committed:
Merge branch 'master' into export-D22640839

mrshenli reviewed (approved):
Thanks for adding this, LGTM!

osandoval-fb committed:
changed wait all to private

osandoval-fb committed:
Merge branch 'master' into export-D22640839
